qiskit_experiments/framework/base_analysis.py

```diff
@@ -32,17 +36,107 @@ class BaseAnalysis(ABC):
+    This method should not have side-effects on the analysis class itself
+    since it could potentially be called asynchronously in multiple threads.
+    Any configurable option values should be specified in the `_default_options`
+    class method. These values can be overriden by a user by calling the
+    `set_options` method or for a single-run can be specified by passing kwarg
+    options to the :meth:`run` method.
     """
 
+    def __init__(self):
+        """Initialize the analysis object."""
+        # Analysis options
+        self._options = self._default_options()
+
+        # Store keys of non-default options
+        self._set_options = set()
+
+    def __new__(cls, *args, **kwargs):
+        """Store init args and kwargs for subclass __init__ methods"""
+        # This method automatically stores all arg and kwargs from subclass
+        # init methods for use in converting an experiment to config
+
+        # Get all non-self init args and kwarg names for subclass
+        spec = inspect.getfullargspec(cls.__init__)
+        init_arg_names = spec.args[1:]
+        num_init_kwargs = len(spec.defaults) if spec.defaults else 0
+        num_init_args = len(init_arg_names) - num_init_kwargs
+
+        # Convert passed values for args and kwargs into an ordered dict
+        # This will sort args passed as kwargs and kwargs passed as
+        # positional args in the function call
+        num_call_args = len(args)
+        ord_args = OrderedDict()
+        ord_kwargs = OrderedDict()
+        for i, argname in enumerate(init_arg_names):
+            if i < num_init_args:
+                update = ord_args
+            else:
+                update = ord_kwargs
+            if i < num_call_args:
+                update[argname] = args[i]
+            elif argname in kwargs:
+                update[argname] = kwargs[argname]
+
+        # pylint: disable = attribute-defined-outside-init
+        instance = super(BaseAnalysis, cls).__new__(cls)
+        instance.__init_args__ = ord_args
+        instance.__init_kwargs__ = ord_kwargs
+        return instance
+
+    @property
+    def config(self) -> AnalysisConfig:
+        """Return the config dataclass for this analysis"""
+        args = tuple(getattr(self, "__init_args__", OrderedDict()).values())
+        kwargs = dict(getattr(self, "__init_kwargs__", OrderedDict()))
+        # Only store non-default valued options
+        options = dict((key, getattr(self._options, key)) for key in self._set_options)
+        return AnalysisConfig(
+            cls=type(self),
+            args=args,
+            kwargs=kwargs,
+            options=options,
+        )
+
+    @classmethod
+    def from_config(cls, config: Union[AnalysisConfig, Dict]) -> "BaseAnalysis":
+        """Initialize an analysis class from analysis config"""
+        if isinstance(config, dict):
+            config = AnalysisConfig(**dict)
+        ret = cls(*config.args, **config.kwargs)
```

Comment:
Seems like you need class check here (same for experiment one). Currently we can initialize an instance from a config object of different class.

Reply:
This is intentional since class in the config could be None or a partially deserialized one. If someone really wants to try and initialize one experiment with config of another they can so long as they have compatible configs.